feat(cache-browser-local-storage): Implemented TTL support to cached items

SanderFlooris opened the pull request:
## Feature summary
This introduces the ability to set a TTL for cached items, in order to reduce the size of this cache and prevents the cache from permanently holding onto the items.

This feature is optional, and if the `timeToLive` option is not configured, it will behave like how it currently does.

## Why?
We ran into storage quotas, and saw that this cache can grow to a couple MB after a while. We made our own patched version, but it would be nice if this feature was implemented upstream.

Providing this fix in the form of a pull request was discussed via the [support.algolia.com/hc/requests/551781](http://support.algolia.com/hc/requests/551781) support ticket.

SanderFlooris commented on packages/cache-browser-local-storage/src/createBrowserLocalStorageCache.ts:
```diff
@@ -19,6 +19,38 @@ export function createBrowserLocalStorageCache(options: BrowserLocalStorageOptio
+
+    if (!timeToLive) {
+      return;
+    }
```
@Haroenv It has been updated 👍

Haroenv replied:
Sorry, I meant right after creating the timeToLive variable, so after line 27, we wouldn't need to do the Object.entries etc.

SanderFlooris replied:
Ah that's what you mean. This was done in order to deal with existing cache items that were made before this PR is merged. 
Line 30 - 34 searches for old cache items and removes them, as they are no longer in a format that is supported.

shortcuts commented on packages/cache-browser-local-storage/src/types/BrowserLocalStorageOptions.ts:
```diff
@@ -4,6 +4,11 @@ export type BrowserLocalStorageOptions = {
+  /**
+   * The time to live for each cached item in seconds.
+   */
+  readonly ttl?: number;
```
I guess we can go full length here and use `timeToLive` to make it easier for everyone, wdyt?

SanderFlooris replied:
Took a look at some of the other `Options.ts` files, and I do agree with this suggestion.
Not only does it make it a bit easier to read, it also makes it more consistent with other configurations.

SanderFlooris replied:
It has been updated 👍

shortcuts reviewed (commented):
good idea, thanks for the contribution!!

shortcuts reviewed (commented):
Thanks for the contribution! Would you mind adding a basic test just to assert that it gets deleted?

SanderFlooris commented:
I've added 2 new tests for this feature, `reads unexpired timeToLive keys` and `deletes expired keys`.
I also updated `creates a namespace within local storage` and `can be cleared`, as they broke with this change.

Since this question might come up, I'll answer it in advance. 
We need to compare if it's an empty object, this is because the `cache.get` that is above it will also run the `removeOutdatedCacheItems` function. 
This will automatically create a new namespace, so for this reason I've moved that specific check to be above the `cache.get`, and added another one at the bottom to see if the namespace is empty.
![Scherm­afbeelding 2023-04-28 om 10 16 16](https://user-images.githubusercontent.com/116890267/235093770-9039cb23-1e8d-4ae8-b4a9-66ba10852127.png)

SanderFlooris commented:
@shortcuts Are there any updates yet regarding this PR? Or is it still being processed internally?

shortcuts commented:
> @shortcuts Are there any updates yet regarding this PR? Or is it still being processed internally?

oh sorry I forgot, checkingt!

shortcuts reviewed (dismissed):
looks good! thanks for the contribution :D I'll port it to the next major as stated here https://github.com/algolia/api-clients-automation/issues/1489

shortcuts commented:
hey, could you bump the bundle size please? https://bundlesize-store.vercel.app/build?info=%7B%22files%22%3A%5B%7B%22maxSize%22%3A8140%2C%22path%22%3A%22packages%2Falgoliasearch%2Fdist%2Falgoliasearch.umd.js%22%2C%22size%22%3A8257%2C%22compression%22%3A%22gzip%22%7D%2C%7B%22maxSize%22%3A4505%2C%22path%22%3A%22packages%2Falgoliasearch%2Fdist%2Falgoliasearch-lite.umd.js%22%2C%22size%22%3A4611%2C%22compression%22%3A%22gzip%22%7D%2C%7B%22maxSize%22%3A4300%2C%22path%22%3A%22packages%2Frecommend%2Fdist%2Frecommend.umd.js%22%2C%22size%22%3A4371%2C%22compression%22%3A%22gzip%22%7D%5D%2C%22repo%22%3A%22algolia%2Falgoliasearch-client-javascript%22%2C%22branch%22%3A%22pull%2F1457%2Fhead%22%2C%22commit_message%22%3A%22%22%2C%22sha%22%3A%228c17ae5386983ee7bea87e940ffb78764f1049b0%22%7D

SanderFlooris commented:
@shortcuts Just making sure, I just have to update the `bundlesize` inside `package.json` right?

shortcuts commented:
> @shortcuts Just making sure, I just have to update the `bundlesize` inside `package.json` right?

exactly!

SanderFlooris commented:
@shortcuts It has been updated :+1:

SanderFlooris commented:
Hmm, seems like the bundle size is a bit bigger then when I locally ran the test.
Currently recompiling everything to see if it will fail the bundle test afterwards.

SanderFlooris commented:
After a fresh compile the size was indeed bigger than before, this has been fixed now

shortcuts reviewed (approved):
Thanks a lot!
